### PR TITLE
Rewrite spec file generation

### PR DESCRIPTION
**Pull request by ekohl**
I ended up refactoring so much that I didn't spend time to pull it into separate commits. Some changes:

* Use handlebars to do templating
* Split into two template files to keep the logic simple
* Set the dependencies for single-style packages
* Use a package to normalize the NPM output rather than do it ourselves

### CODE REVIEW

File: bundle.mustache [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+%prep
+mkdir -p %{npm_cache_dir}
+for tgz in %{sources}; do
+  echo $tgz | grep -q registry.npmjs.org || npm cache add --cache %{npm_cache_dir} $tgz
+done
+%setup -T -q -a {{{PROVIDES.length}}} -D -n %{npm_cache_dir}
+
+%build
+npm install --cache-min Infinity --cache %{npm_cache_dir} --no-optional --global-style true %{npm_name}@%{version}
```
Comment (dLobatog) [inline bundle.mustache]: Could you add `--no-shrinkwrap` here too?

File: bundle.mustache [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+{{/each}}
+{{#each LICENSEFILES}}
+%license {{{.}}}
+{{/each}}
+{{#each DOCFILES}}
+%doc {{{.}}}
+{{/each}}
+
+%changelog
+* {{{DATE}}} author@example.com <Author Name> - {{{VERSION}}}-{{{RELEASE}}}
```
Comment (dLobatog) [inline bundle.mustache]: Maybe worth keeping it empty since we don't use this entry at all with tito

File: lib/spec_file_generator.js [hunk 3]
```diff
@@ -1,163 +1,116 @@
...
-function listBinaryFiles(binaries) {
-  if (typeof(binaries) === 'string') {
-    return '%{_bindir}/' + getBasename(binaries);
-  } else {
-    var binary_files = '';
-    for(var binary in binaries) {
-      binary_files += '%{_bindir}/' + getBasename(binaries[binary]);
-      binary_files += '\n';
+  var sortedDeps = []
+  for (var dep in deps) {
```
Comment (dLobatog) [inline lib/spec_file_generator.js]: nitpick, just `dep in deps`

File: lib/spec_file_generator.js [hunk 4]
```diff
@@ -1,163 +1,116 @@
...
-}
-
-function sourcesToCache(deps) {
-  return 'mkdir -p %{npm_cache_dir}\n\
-for tgz in %{sources}; do\n\
-  echo $tgz | grep -q registry.npmjs.org || npm cache add --cache %{npm_cache_dir} $tgz\n\
-done\n\
-\n\
-%setup -T -q -a ' + deps.length + ' -D -n %{npm_cache_dir}';
+  var template = Handlebars.compile(spec_file);
```
Comment (dLobatog) [inline lib/spec_file_generator.js]: `template` is is already defined in the function signature (it's used in the line above this even), could you rename it?

**Review by dLobatog — CHANGES_REQUESTED** [review]
@ekohl I only have a few comments, overall I think it's a great improvement. Do you have any plans to use other templates for specs that are not the ones included? 

I tested with a few of the packages in foreman-packaging which were built just fine. I'm happy to just release this (after fixing the comments in here) and iterate quickly to fix any further problems.

File: bundle.mustache (hunk 2 above)
Comment (ekohl) [reply bundle.mustache]: Looks like I did that locally but didn't push it

**Comment by ekohl** [comment]
Updated. It does a bunch more with file handling and ignoring certain files. Please have a good look because I might have forgotten this or edge cases.

**Review by dLobatog — APPROVED** [review]
I've used this to create a few specs already - no issues for the moment. I'll merge and release, if anything is broken we can always do a patch release.

Thanks @ekohl

**Pull request merged**